### PR TITLE
UCT/IB - added support for RoCE.

### PR DESCRIPTION
**Pull request by alinask**
currently only for the ud (verbs) transport.

related to https://github.com/openucx/ucx/issues/193

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1571/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2238/ for details.

File: src/uct/ib/base/ib_device.h [hunk 1]
```diff
@@ -48,15 +49,18 @@ enum {
...
-    UCT_IB_ADDRESS_SCOPE_LINK_LOCAL,   /* Subnet-local address */
-    UCT_IB_ADDRESS_SCOPE_SITE_LOCAL,   /* Site local, 16-bit subnet prefix */
-    UCT_IB_ADDRESS_SCOPE_GLOBAL        /* Global, 64-bit subnet prefix */
-} uct_ib_address_scope_t;
+    UCT_IB_ADDRESS_TYPE_LINK_LOCAL,   /* Subnet-local address */
+    UCT_IB_ADDRESS_TYPE_SITE_LOCAL,   /* Site local, 16-bit subnet prefix */
+    UCT_IB_ADDRESS_TYPE_GLOBAL,       /* Global, 64-bit subnet prefix */
+    UCT_IB_ADDRESS_TYPE_ETH_LOCAL,    /* RoCE - Subnet-local address */
+    UCT_IB_ADDRESS_TYPE_ETH_GLOBAL    /* RoCE - Subnet-global address */
+} uct_ib_address_type_t;
```
Comment (shssf) [inline src/uct/ib/base/ib_device.h]: As I understand, the order if the enum elements must be unchanged. Overwise, the handling algorithm will be broken. Perhaps, it might be helpful to add warning about this.

Comment (alinask) [inline src/uct/ib/base/ib_device.h]: I only added two values to the end of the enum. the previous values remain the same.
or did you mean something else?

Comment (alinask) [inline src/uct/ib/base/ib_device.h]: ok no problem, i'll add a commnet.

**Comment by alinask** [comment]
@yosefe please review as well.

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1703,9 +1703,9 @@ void ucp_dt_destroy(ucp_datatype_t datatype);
  * subsequent communication operations to the same @ref ucp_worker_h "worker"
  * which follow the call to @ref ucp_worker_fence "fence".
  *
- * @note The primary diference between @ref ucp_worker_fence "ucp_worker_fence()"
+ * @note The primary difference between @ref ucp_worker_fence "ucp_worker_fence()"
  * and the @ref ucp_worker_flush "ucp_worker_flush()" is the fact the fence
- * routine does not gurantee completion of the operations on the call return but
+ * routine does not guarantee completion of the operations on the call return but
  * only ensures the order between communication operations. The
  * @ref ucp_worker_flush "flush" operation on return guarantees that all
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Thanks for fixing this, but I would prefer these changes to go as a separate PR, since it has nothing to do with RoCE.

File: src/uct/ib/base/ib_device.c [hunk 3]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
-        return UCT_IB_ADDRESS_SCOPE_SITE_LOCAL;
+        return UCT_IB_ADDRESS_TYPE_SITE_LOCAL;
     } else {
-        return UCT_IB_ADDRESS_SCOPE_GLOBAL;
+        return UCT_IB_ADDRESS_TYPE_GLOBAL;
     }
 }
 
-size_t uct_ib_address_size(uct_ib_address_scope_t scope)
+size_t uct_ib_address_size(uct_ib_address_type_t type)
```
Comment (shamisp) [inline src/uct/ib/base/ib_device.c]: I will make another side note. It is easier to review the code when renaming, actual change, and example/test are separates to three different commits with the same PR.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2239/ for details.

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (alinask) [inline src/ucp/api/ucp.h]: I see, it's just that it was a small typo..
let's keep this one and i'll do as you asked from now on.

File: src/uct/ib/base/ib_device.h [hunk 4]
```diff
@@ -48,15 +49,18 @@ enum {
...
 typedef enum {
-    UCT_IB_ADDRESS_SCOPE_LINK_LOCAL,   /* Subnet-local address */
-    UCT_IB_ADDRESS_SCOPE_SITE_LOCAL,   /* Site local, 16-bit subnet prefix */
-    UCT_IB_ADDRESS_SCOPE_GLOBAL        /* Global, 64-bit subnet prefix */
-} uct_ib_address_scope_t;
+    UCT_IB_ADDRESS_TYPE_LINK_LOCAL,   /* Subnet-local address */
+    UCT_IB_ADDRESS_TYPE_SITE_LOCAL,   /* Site local, 16-bit subnet prefix */
+    UCT_IB_ADDRESS_TYPE_GLOBAL,       /* Global, 64-bit subnet prefix */
+    UCT_IB_ADDRESS_TYPE_ETH_LOCAL,    /* RoCE - Subnet-local address */
+    UCT_IB_ADDRESS_TYPE_ETH_GLOBAL    /* RoCE - Subnet-global address */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: Think we should remove UCT_IB_ADDRESS_TYPE_ETH_GLOBAL/LOCAL and keep only UCT_IB_ADDRESS_TYPE_ETH

File: src/uct/ib/base/ib_device.c [hunk 5]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
-    case UCT_IB_ADDRESS_SCOPE_GLOBAL:
+    case UCT_IB_ADDRESS_TYPE_GLOBAL:
         return sizeof(uct_ib_address_t) +
                sizeof(uint16_t) + /* lid */
                sizeof(uint64_t) + /* if_id */
-               sizeof(uint64_t);  /* subney64 */
+               sizeof(uint64_t);  /* subnet64 */
+    case UCT_IB_ADDRESS_TYPE_ETH_LOCAL:
+        return sizeof(uct_ib_address_t) +
+               16 * sizeof(uint8_t);  /* raw gid */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: sizeof(union ibv_gid)

File: src/uct/ib/base/ib_device.c [hunk 6]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
-
-        if (scope >= UCT_IB_ADDRESS_SCOPE_GLOBAL) {
-            /* Global */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET64;
-            *(uint64_t*)ptr = gid->global.subnet_prefix;
-        } else {
-            /* Site-local */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
-            *(uint16_t*)ptr = gid->global.subnet_prefix >> 48;
+    if (type < UCT_IB_ADDRESS_TYPE_ETH_LOCAL) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: let's not assume order of enum here.
if (type == ETH) {
  // eth code
} else {
  /* IB */
}

File: src/uct/ib/base/ib_device.c [hunk 7]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
 void uct_ib_address_unpack(const uct_ib_address_t *ib_addr, uint16_t *lid,
                            uint8_t *is_global, union ibv_gid *gid)
 {
     const void *ptr = ib_addr + 1;
 
-    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_LID) {
-        *lid = *(uint16_t*)ptr;
-        ptr += sizeof(uint16_t);
-    } else {
+    if (ucs_unlikely(ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: no need for unlikely, this is not data path

File: src/uct/ib/base/ib_device.c [hunk 8]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
+        /* RoCE */
+        *is_global = 1;
         *lid = 0;
-    }
-
-    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_IF_ID) {
-        gid->global.interface_id = *(uint64_t*)ptr;
-        ptr += sizeof(uint64_t);
+        memcpy(gid->raw, ptr, 16 * sizeof(uint8_t));    /* uint8_t raw[16]; */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: sizeof(gid->raw)

File: src/uct/ib/base/ib_device.c [hunk 9]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
     } else {
-        gid->global.interface_id = 0;
-    }
+        /* IB */
 
-    gid->global.subnet_prefix = UCT_IB_LINK_LOCAL_PREFIX; /* Default prefix */
-    *is_global                = 0;
+        if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_LID) {
+            *lid = *(uint16_t*) ptr;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: code style - redundant space

File: src/uct/ib/base/ib_device.c [hunk 10]
```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
...
-    }
-
-    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_IF_ID) {
-        gid->global.interface_id = *(uint64_t*)ptr;
-        ptr += sizeof(uint64_t);
+        memcpy(gid->raw, ptr, 16 * sizeof(uint8_t));    /* uint8_t raw[16]; */
     } else {
-        gid->global.interface_id = 0;
-    }
+        /* IB */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: no need to put all this in 'else', just need to unpack dgid-raw if exists

File: src/uct/ib/base/ib_device.h [hunk 11]
```diff
@@ -48,15 +49,18 @@ enum {
 enum {
     UCT_IB_DEVICE_FLAG_MLX4_PRM = UCS_BIT(1),   /* Device supports mlx4 PRM */
     UCT_IB_DEVICE_FLAG_MLX5_PRM = UCS_BIT(2),   /* Device supports mlx5 PRM */
-    UCT_IB_DEVICE_FLAG_DC       = UCS_BIT(3)    /* Device supports DC */
+    UCT_IB_DEVICE_FLAG_DC       = UCS_BIT(3),   /* Device supports DC */
+    UCT_IB_DEVICE_FLAG_ONLY_IB  = UCS_BIT(4)    /* Require only IB */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: UCT_IB_DEVICE_FLAG_LINK_IB

File: src/uct/ib/base/ib_device.h [hunk 12]
```diff
@@ -66,7 +70,8 @@ enum {
     UCT_IB_ADDRESS_FLAG_LID      = UCS_BIT(0),
     UCT_IB_ADDRESS_FLAG_IF_ID    = UCS_BIT(1),
     UCT_IB_ADDRESS_FLAG_SUBNET16 = UCS_BIT(2),
-    UCT_IB_ADDRESS_FLAG_SUBNET64 = UCS_BIT(3)
+    UCT_IB_ADDRESS_FLAG_SUBNET64 = UCS_BIT(3),
+    UCT_IB_ADDRESS_FLAG_GID_RAW  = UCS_BIT(4)
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: UCT_IB_ADDRESS_FLAG_GID ("raw" is only when it's inside the union. the correct terminology is just a "gid")

File: src/uct/ib/base/ib_device.h [hunk 13]
```diff
@@ -75,11 +80,13 @@ enum {
  */
 typedef struct uct_ib_address {
     uint8_t            flags;
-    /* Following fields appear in this order (if specified by flags) :
+    /* Following fields appear in this order (if specified by flags) unless RoCE is used:
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: we can just say that the full gid always appears last, as specified by flags. no need to say it's only for RoCE / IB.

File: src/uct/ib/base/ib_device.h [hunk 14]
```diff
@@ -75,11 +80,13 @@ enum {
...
      * - uint16_t lid
      * - uint64_t if_id
      * - uint16_t subnet16
      * - uint64_t subnet64
+     * For RoCE:
+     * - uint8_t raw
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: this is uint8_t gid[16]

File: src/uct/ib/base/ib_iface.c [hunk 15]
```diff
@@ -177,7 +177,22 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface, const uct_device_addr_
     uint16_t lid;
 
     uct_ib_address_unpack(ib_addr, &lid, &is_global, &gid);
-    return (gid.global.subnet_prefix == iface->gid.global.subnet_prefix);
+
+    if ((ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW) == 0) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: 1. !(ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW)
2. need to check that the local link layer type is matching the remote one

File: src/uct/ib/base/ib_iface.c [hunk 16]
```diff
@@ -177,7 +177,22 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface, const uct_device_addr_
...
+        /* IB */
+        return (gid.global.subnet_prefix == iface->gid.global.subnet_prefix);
+    } else {
+        /* RoCE */
+        /* check that there is no lid and that the raw_gid is on */
+        if ((ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW) &&
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: why check ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW again, if it was already checked in the previous 'if'?

Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: also, instead of doing 
if (cond) {
 return 1
} else {
 return 0
}

do just: 
return cond

File: src/uct/ib/base/ib_iface.c [hunk 17]
```diff
@@ -444,24 +471,22 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
         goto err_destroy_send_cq;
     }
 
-    if (!uct_ib_device_is_port_ib(dev, self->port_num)) {
-        ucs_error("Unsupported link layer");
-        status = UCS_ERR_UNSUPPORTED;
-        goto err_destroy_recv_cq;
+    /* Address scope and size */
+    self->addr_type = uct_ib_address_scope(self->gid.global.subnet_prefix);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: we should not call uct_ib_address_scope() at all when link layer is Ethernet

File: src/uct/ib/base/ib_iface.c [hunk 18]
```diff
@@ -602,8 +623,21 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
     /* Calculate packet overhead  */
     mtu                   = ucs_min(uct_ib_mtu_value(active_mtu),
                                     iface->config.seg_size);
-    extra_pkt_len         = UCT_IB_LRH_LEN + UCT_IB_BTH_LEN + xport_hdr_len +
-                            UCT_IB_ICRC_LEN + UCT_IB_VCRC_LEN + UCT_IB_DELIM_LEN;
+
+#if HAVE_DECL_IBV_LINK_LAYER_INFINIBAND
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: why node #if HAVE_DECL_IBV_LINK_LAYER_ETHERNET?

File: src/uct/ib/base/ib_iface.c [hunk 19]
```diff
@@ -602,8 +623,21 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
...
+    if (uct_ib_iface_port_attr(iface)->link_layer != IBV_LINK_LAYER_ETHERNET) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: would be nicer to handle do if (==)  than if (!=)  - reverse order of if/else

File: src/uct/ib/base/ib_iface.c [hunk 20]
```diff
@@ -602,8 +623,21 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
...
+        extra_pkt_len = UCT_IB_LRH_LEN + UCT_IB_BTH_LEN + xport_hdr_len +
+                        UCT_IB_ICRC_LEN + UCT_IB_VCRC_LEN + UCT_IB_DELIM_LEN;
+    } else {
+        extra_pkt_len = UCT_IB_GRH_LEN + UCT_IB_ROCE_LEN + UCT_IB_BTH_LEN + xport_hdr_len +
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: also - pls try to avoid code duplication here. For example, only UCT_IB_LRH_LEN or UCT_IB_GRH_LEN + UCT_IB_ROCE_LEN should be inside the conditionals. the rest should be outside the 'if'

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (shamisp) [inline src/ucp/api/ucp.h]: Ok, but now it has to be approved by all maintainers, even so it is  local IB modification.

Comment (alinask) [inline src/ucp/api/ucp.h]: Ok, will remove.

File: src/uct/ib/base/ib_device.c (hunk 10 above)
Comment (alinask) [inline src/uct/ib/base/ib_device.c]: but it always exist.

File: src/uct/ib/base/ib_device.h (hunk 13 above)
Comment (alinask) [inline src/uct/ib/base/ib_device.h]: ok, i'll keep both, to make it more clear.

File: src/uct/ib/base/ib_iface.c (hunk 15 above)
Comment (alinask) [inline src/uct/ib/base/ib_iface.c]: how can i tell the remote link layer? it's not passed in the address currenly.

File: src/uct/ib/base/ib_device.c (hunk 10 above)
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: it is not present in the packed address unless the proper flag (which you check in the 'if') is set

File: src/uct/ib/base/ib_iface.c (hunk 15 above)
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: may add a flag to address flags which means LINK_LAYER_IB

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2248/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1580/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2249/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1581/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2250/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2251/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2252/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2253/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2254/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2255/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2256/ for details.

File: src/uct/ib/base/ib_device.c [hunk 21]
```diff
@@ -340,99 +340,121 @@ uint8_t uct_ib_to_fabric_time(double time)
...
-    case UCT_IB_ADDRESS_SCOPE_GLOBAL:
+    case UCT_IB_ADDRESS_TYPE_GLOBAL:
         return sizeof(uct_ib_address_t) +
                sizeof(uint16_t) + /* lid */
                sizeof(uint64_t) + /* if_id */
-               sizeof(uint64_t);  /* subney64 */
+               sizeof(uint64_t);  /* subnet64 */
+    case UCT_IB_ADDRESS_TYPE_ETH:
+        return sizeof(uct_ib_address_t) +
+               sizeof(union ibv_gid) * sizeof(uint8_t);  /* raw gid */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: no need for  `* sizeof(uint8_t)`

File: src/uct/ib/base/ib_device.c [hunk 22]
```diff
@@ -340,99 +340,121 @@ uint8_t uct_ib_to_fabric_time(double time)
...
     void *ptr = ib_addr + 1;
 
     ib_addr->flags  = 0;
 
-    /* LID */
-    /* TODO check IB link layer of the port */
-    ib_addr->flags |= UCT_IB_ADDRESS_FLAG_LID;
-    *(uint16_t*)ptr = lid;
-    ptr += sizeof(uint16_t);
+    if (type  !=  UCT_IB_ADDRESS_TYPE_ETH) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: 2 spaces

File: src/uct/ib/base/ib_device.c [hunk 23]
```diff
@@ -340,99 +340,121 @@ uint8_t uct_ib_to_fabric_time(double time)
...
+                /* Site-local */
+                ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
+                *(uint16_t*) ptr = gid->global.subnet_prefix >> 48;
+            }
         }
+    } else {
+        /* RoCE */
+        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_LINK_LAYER_ETH;
+        /* in this case we don't use the lid */
+        ib_addr->flags &= ~UCT_IB_ADDRESS_FLAG_LID;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: no need to remove LID flags because it was not added

File: src/uct/ib/base/ib_device.h [hunk 24]
```diff
@@ -19,6 +19,7 @@
 #define UCT_IB_LRH_LEN              8   /* IB Local routing header */
 #define UCT_IB_GRH_LEN              40  /* IB GLobal routing header */
 #define UCT_IB_BTH_LEN              12  /* IB base transport header */
+#define UCT_IB_ROCE_LEN             14  /* Ethernet header - 6B for Destination MAC + 6B for Source MAC + 2B Type (RoCE) */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: line longer than 80

File: src/uct/ib/base/ib_device.c [hunk 25]
```diff
@@ -340,99 +340,121 @@ uint8_t uct_ib_to_fabric_time(double time)
...
+        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_GID;
+        memcpy(ptr, gid->raw, sizeof(gid->raw) * sizeof(uint8_t));    /* uint8_t raw[16]; */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: line longer than 80

File: src/uct/ib/base/ib_device.h [hunk 26]
```diff
@@ -75,11 +81,13 @@ enum {
  */
 typedef struct uct_ib_address {
     uint8_t            flags;
-    /* Following fields appear in this order (if specified by flags) :
+    /* Following fields appear in this order (if specified by flags) . The full gid always appears last:
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: line longer than 80

File: src/uct/ib/base/ib_iface.c [hunk 27]
```diff
@@ -177,7 +177,27 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface, const uct_device_addr_
     uint16_t lid;
 
     uct_ib_address_unpack(ib_addr, &lid, &is_global, &gid);
-    return (gid.global.subnet_prefix == iface->gid.global.subnet_prefix);
+
+#if HAVE_DECL_IBV_LINK_LAYER_INFINIBAND
+    if (!(ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID) ) {
+        /* IB */
+        return ((uct_ib_iface_port_attr(iface)->link_layer == IBV_LINK_LAYER_INFINIBAND) &&
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: long line

File: src/uct/ib/base/ib_iface.c [hunk 28]
```diff
@@ -177,7 +177,27 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface, const uct_device_addr_
     uint16_t lid;
 
     uct_ib_address_unpack(ib_addr, &lid, &is_global, &gid);
-    return (gid.global.subnet_prefix == iface->gid.global.subnet_prefix);
+
+#if HAVE_DECL_IBV_LINK_LAYER_INFINIBAND
+    if (!(ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID) ) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: need to check for UCT_IB_ADDRESS_FLAG_LINK_LAYER_IB 1st

Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: can be done much simpler:
#ifdef HAVE_ ... _ETHERNET
if (link_layer == ETH) {
   return flags & ETH;
} else if (link_layer == IB) 
#endif
{
 return (flags & IB) && (gid.subnet_prefix == iface->subnet_prefix)
}

File: src/uct/ib/base/ib_iface.c [hunk 29]
```diff
@@ -272,6 +292,20 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
     return UCS_ERR_INVALID_PARAM;
 }
 
+#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
+static int uct_ib_iface_is_gid_raw_empty(uint8_t *gid_raw)
+{
+    int i;
+
+    for (i = 0; i < 16; i++) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: simpler and more efficient: 
return (*(uint64_t*)gid_raw == 0) && (*(uint64_t*)(gid_raw + 8) == 0)

File: src/uct/ib/base/ib_iface.c [hunk 30]
```diff
@@ -444,24 +490,25 @@ UCS_CLASS_INIT_FUNC(uct_ib_iface_t, uct_ib_iface_ops_t *ops, uct_md_h md,
...
-        status = UCS_ERR_UNSUPPORTED;
-        goto err_destroy_recv_cq;
-    }
-
     /* Address scope and size */
-    self->addr_scope = uct_ib_address_scope(self->gid.global.subnet_prefix);
-    self->addr_size  = uct_ib_address_size(self->addr_scope);
+#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
+    if (uct_ib_iface_port_attr(self)->link_layer == IBV_LINK_LAYER_ETHERNET) {
+           self->addr_type = UCT_IB_ADDRESS_TYPE_ETH;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: indentation

File: src/uct/ib/base/ib_iface.c [hunk 31]
```diff
@@ -602,8 +645,19 @@ ucs_status_t uct_ib_iface_query(uct_ib_iface_t *iface, size_t xport_hdr_len,
...
-                            UCT_IB_ICRC_LEN + UCT_IB_VCRC_LEN + UCT_IB_DELIM_LEN;
+
+    extra_pkt_len = UCT_IB_BTH_LEN + xport_hdr_len +  UCT_IB_ICRC_LEN + UCT_IB_VCRC_LEN + UCT_IB_DELIM_LEN;
+
+#if HAVE_DECL_IBV_LINK_LAYER_ETHERNET
+    if (uct_ib_iface_port_attr(iface)->link_layer == IBV_LINK_LAYER_ETHERNET) {
+        extra_pkt_len += UCT_IB_GRH_LEN + UCT_IB_ROCE_LEN;
+            /* TODO check if UCT_IB_DELIM_LEN is present in RoCE as well */
+    } else {
+                extra_pkt_len += UCT_IB_LRH_LEN;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c]: 1. indentation
2. pls unite the code of else with the code of #else

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 32]
```diff
@@ -715,7 +715,7 @@ uct_ud_mlx5_query_resources(uct_md_h md,
     ucs_trace_func("");
     /* TODO take transport overhead into account */
     return uct_ib_device_query_tl_resources(&ucs_derived_of(md, uct_ib_md_t)->dev,
-                                            "ud_mlx5", UCT_IB_DEVICE_FLAG_MLX5_PRM,
+                                            "ud_mlx5", UCT_IB_DEVICE_FLAG_MLX5_PRM | UCT_IB_DEVICE_FLAG_LINK_IB,
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: pls split the line

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2270/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1597/ for details (Mellanox internal link).

**Comment by alinask** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2271/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1598/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:
